### ISSUE TITLE
MoltenVK Enhancement Priorities

### ISSUE THREAD

**Issue by billhollings** [KhronosGroup/MoltenVK#1975]
This list provides a place to consolidate the priorities and status of some of the larger MoltenVK enhancements whose development is either underway, or will be soon. This list will also help to prevent multiple MoltenVK devs unknowingly working independently on significant enhancements. 

If you are, or soon plan to be, working on a significant MoltenVK enhancement, please comment below, and we'll add it to the list (not every enhancement needs to go on this list, if you are banging out a small enhancement). If you want to take over one of the un-started enhancements, let us know below and it can be assigned to you.

If you are waiting on a significant enhancement, and want to request that it be added to the list, please request that in the issue for that enhancement.

The list is in order of attack. This list does not include bug fixing work.

| | In Dev | Merged | Developer | Enhancement| Deps | Issue# | PR# |
| :-: | :-: | :-: |:-- |:--  |:-: |:-: |:-: 
| 1 | :white_check_mark: |  | @billhollings  | Metal 3 Arg Buffs (enhancing `VK_EXT_descriptor_indexing`) | | | |
| 2 | :white_check_mark: |  | @AntarticCoder  | Vulkan Acceleration Structures | | #1956 | #1967 |
| 3 | :white_check_mark: |  | @gpx1000 | Transform Feedback | | | #1943 |
| 4 |  |  | @billhollings  | (macOS 14+ only) Integrate Apple's Shader Converter tech (including Geom & Tess shaders) | 1 |  | |
| 5 |  |  | @billhollings  | `VK_KHR_synchronization2` | | #1514 | |
| 6 |  |  | @billhollings  | `VK_EXT_descriptor_buffer` | 1, 5 | #1776 | |
| 7 | :white_check_mark: |  | @billhollings  | Pipeline caching using `MTLBinaryArchives` | | #1765 | |

**Comment by AntarticCoder** [comment]
If anyone's interested in implementing this, the spec is [here](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_maintenance4.html). `VK_KHR_maintenance4` is also considered core in Vulkan 1.3, so implementing it is almost a no brainer.

**Comment by billhollings** [comment]
> VK_KHR_maintenance4 is also considered core in Vulkan 1.3

Yes. Vulkan 1.3 requirements are being tracked separately in issue #1930. I've clarified that above, along with how this list interacts with that list.

**Comment by oscarbg** [comment]
Hi,
Seems geo and tf are part of the present crossover 23 beta1 release, which says, 
“We’re pleased to announce that this release also includes initial support for geometry shaders and transform feedback on macOS”..

I’m pretty curious to see how this geo shader implementation compares to “Apple's Shader Converter tech” both performance wise and compatibility wise once the two ara available on MoltenVK.. at least on MacOS 14+..
Exciting times..

**Comment by AndreaCicca** [comment]
> Hi, Seems geo and tf are part of the present crossover 23 beta1 release, which says, “We’re pleased to announce that this release also includes initial support for geometry shaders and transform feedback on macOS”..
> 
> I’m pretty curious to see how this geo shader implementation compares to “Apple's Shader Converter tech” both performance wise and compatibility wise once the two ara available on MoltenVK.. at least on MacOS 14+.. Exciting times..

@js6i has a PR ([#1815](https://github.com/KhronosGroup/MoltenVK/pull/1815)) where he said that "Currently we're focused on preparing the next CrossOver release, there will be updates after it's out."

I don't know what will be the plan for that PR if the project want to use Apple's Shader Converter for Geometry shader

**Comment by Zingam** [comment]
`VK_EXT_shader_object` looks like an extension that should be included sooner rather than later.
https://www.khronos.org/blog/you-can-use-vulkan-without-pipelines-today

**Comment by K0bin** [comment]
> Metal 3 Arg Buffs (enhancing bindless resources)

It would be very nice if MoltenVK could also properly support "aliasing descriptors" with the reworked binding code.

The shader compiler really hates this pattern right now:
```glsl
layout(set = 0, binding = 0) uniform sampler2D s0_2d;
layout(set = 0, binding = 0) uniform sampler3D s0_3d;
```

**Comment by tgcordell** [comment]
I believe Enhancement #3 above (Wide lines and basic support for private APIs) is complete and merged, but the table has not been updated to reflect this. Are there additional updates to the table that can be made to track progress?

**Comment by billhollings** [comment]
> I believe Enhancement #3 above (Wide lines and basic support for private APIs) is complete and merged, but the table has not been updated to reflect this. Are there additional updates to the table that can be made to track progress?

Yes, the functionality is merged. It's indicated in the checkbox in the _**Merged**_ column for that enhancement.

**Comment by cdavis5e** [comment]
> > I believe Enhancement #3 above (Wide lines and basic support for private APIs) is complete and merged, but the table has not been updated to reflect this. Are there additional updates to the table that can be made to track progress?
> 
> Yes, the functionality is merged. It's indicated in the checkbox in the _**Merged**_ column for that enhancement.

That's because, after I saw @tgcordell's comment, I went and silently updated it.

**Comment by tgcordell** [comment]
I’m most interested in #6, the Metal 3 Arg Buffers. Is there a PR where I can follow the development of that enchantment?

**Comment by billhollings** [comment]
> I’m most interested in #6, the Metal 3 Arg Buffers. Is there a PR where I can follow the development of that enchantment?

Initial work has been done, but there is no active PR yet. We're trying to prioritize it for sometime this summer.

**Comment by orosbogdan** [comment]
Are geometry shaders and tesselation shaders supported on both ios and mac at the moment using moltenvk?

**Comment by billhollings** [comment]
> Are geometry shaders and tesselation shaders supported on both ios and mac at the moment using moltenvk?

Tessellation shaders are supported.

Geometry shaders are awaiting completion of PR #1815, which has stalled due to lack of activity, contributions, and/or prioritized funding support for that feature.

**Comment by Mohamed-S-Badawy** [comment]
Hey, thanks for all the hard work!
Will the descriptor buffer extension come in anytime soon?

**Comment by spnda** [comment]
> Will the descriptor buffer extension come in anytime soon?

I worked on it like a year ago and iirc some specific part of the Vulkan API is not implementable over Metal, and I think most of it was a lot of gymnastics to get it work, since Metal requires us to specify which resources are used in a pass to make them resident. However, given the recent addition of residency set it might be a lot simpler to implement using that, though I don't have the time to look into that at all. See #1776 for more info.
